extensions/amp-story/1.0/amp-story-embedded-component.js

```diff
@@ -0,0 +1,606 @@
+ *   buttonLeft: !Element,
+ *   buttonRight: !Element,
+ *   arrow: !Element,
+ * }}
+ */
+let tooltipElementsDef;
+
+const TAG = 'amp-story-embedded-component';
+
+const ComponentState = {
+  HIDDEN: 0, // Component is present in page, but hasn't been interacted with.
+  FOCUSED: 1, // Component has been clicked, a tooltip should be shown.
+  EXPANDED: 2, // Component is in expanded mode.
+};
+
+/**
+ * Embedded components found in amp-story.
+ */
+export class AmpStoryEmbeddedComponent {
+  /**
+   * @param {!Window} win
+   * @param {!Element} storyEl
+   */
+  constructor(win, storyEl) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Element} */
+    this.storyEl_ = storyEl;
+
+    /** @private {?Element} */
+    this.shadowRoot_ = null;
+
+    /** @private {?Element} */
+    this.focusedStateOverlay_ = null;
+
+    /** @private {?Element} */
+    this.tooltip_ = null;
+
+    /** @private {?Element} */
+    this.tooltipArrow_ = null;
+
+    /** @private @const {!./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = getStoreService(this.win_);
+
+    /** @private @const {!../../../src/service/resources-impl.Resources} */
+    this.resources_ = Services.resourcesForDoc(getAmpdoc(this.win_.document));
+
+    /** @private {?Element} */
+    this.expandedViewOverlay_ = null;
+
+    /**
+     * Target producing the tooltip. Used to avoid building the same
+     * element twice.
+     * @private {?Element} */
+    this.previousTarget_ = null;
+
+    /** @private */
+    this.expandComponentHandler_ = this.onExpandComponent_.bind(this);
+
+    this.storeService_.subscribe(StateProperty.EMBEDDED_COMPONENT, target => {
+      this.onEmbeddedComponentUpdate_(target);
+    });
+
+    this.state_ = ComponentState.HIDDEN;
+  }
+
+  /**
+   * Reacts to embedded component updates.
+   * @param {?Element} target
+   */
+  onEmbeddedComponentUpdate_(target) {
+    switch (this.state_) {
+      case ComponentState.FOCUSED:
+      case ComponentState.HIDDEN:
+        if (target) {
+          // Target is hidden and going into focused mode.
+          this.state_ = ComponentState.FOCUSED;
+
+        } else {
+          // Target is focused and click was outside the tooltip.
+          this.state_ = ComponentState.HIDDEN;
+        }
+        this.onFocusedStateUpdate_(target);
+        break;
+      case ComponentState.EXPANDED:
+        if (target && matches(target,
+            embeddedComponentSelectors().EXPANDED_VIEW_OVERLAY)) {
+          if (matches(target, '.i-amphtml-expanded-view-close-button')) {
+            // Target is expanded and going into hidden mode.
+            this.state_ = ComponentState.HIDDEN;
+            this.toggleExpandedView_(null);
+            this.tooltip_.removeEventListener('click',
+                this.expandComponentHandler_, true);
+          }
+        }
+        else {
+          this.onFocusedStateUpdate_(null);
+          this.toggleExpandedView_(target);
+        }
```

Comment:
I wrote a super long comment trying to simplify these nested if/else statements, but actually I think what's confusing is that the state is `EXPANDED` while the component is either expanded, or not.

An easy way to work around this issue and make this code a lot easier would be having a `setState(state, target)` method, or anything that would work like that.
The goal being always having the UI and the state sync'ed.

```
setState(state, target) {
  switch (state) {
    case ComponentState.FOCUSED:
      this.state = state;
      // Trigger the tooltip UI.
      break;
    case ComponentState.EXPANDED:
      this.state = state;
      // Trigger the expanded UI.
      break;
    ...
  }
}
```
It shouldn't be much of a code change, but should be a lot easier to read. Wdyt?

Reply:
Done